web/concrete/src/Attribute/Key/CollectionKey.php

```diff
@@ -38,104 +41,122 @@ public static function getAttributes($cID, $cvID, $method = 'getValue') {
-			'cID' => $nvc->getCollectionID(),
-			'cvID' => $nvc->getVersionID(),
-			'akID' => $this->getAttributeKeyID(),
-			'avID' => $av->getAttributeValueID()
-		), array('cID', 'cvID', 'akID'));
-		unset($av);
-	}
-
-	public static function add($at, $args, $pkg = false) {
-
-		// legacy check
-		$fargs = func_get_args();
-		if (count($fargs) >= 5) {
-			$at = $fargs[4];
-			$pkg = false;
-			$args = array('akHandle' => $fargs[0], 'akName' => $fargs[1], 'akIsSearchable' => $fargs[2]);
-		}
-
-		CacheLocal::delete('collection_attribute_key_by_handle', $args['akHandle']);
-
-		$ak = parent::add('collection', $at, $args, $pkg);
-		return $ak;
-	}
-
-	public function delete() {
-		parent::delete();
-		$db = Loader::db();
-		$r = $db->Execute('select avID from CollectionAttributeValues where akID = ?', array($this->getAttributeKeyID()));
-		while ($row = $r->FetchRow()) {
-			$db->Execute('delete from AttributeValues where avID = ?', array($row['avID']));
-		}
-		$db->Execute('delete from CollectionAttributeValues where akID = ?', array($this->getAttributeKeyID()));
-	}
+    public static function getColumnHeaderList()
+    {
+        return parent::getList('collection', array('akIsColumnHeader' => 1));
+    }
+
+    public static function getSearchableIndexedList()
+    {
+        return parent::getList('collection', array('akIsSearchableIndexed' => 1));
+    }
+
+    public static function getSearchableList()
+    {
+        return parent::getList('collection', array('akIsSearchable' => 1));
+    }
+
+    public function getAttributeValue($avID, $method = 'getValue')
+    {
+        $av = CollectionAttributeValue::getByID($avID);
+        if (is_object($av)) {
+            $av->setAttributeKey($this);
+            $value = $av->{$method}();
+            $av->__destruct();
+            unset($av);
+            return $value;
+        }
+    }
+
+    public static function getByID($akID)
+    {
+        $ak = new static();
+        $ak->load($akID);
+        if ($ak->getAttributeKeyID() > 0) {
+            return $ak;
+        }
+    }
+
+    public static function getByHandle($akHandle)
+    {
+        $ak = CacheLocal::getEntry('collection_attribute_key_by_handle', $akHandle);
+        if (is_object($ak)) {
+            return $ak;
+        } else if ($ak == -1) {
+            return false;
+        }
+
+        $ak = new static();
+        $ak->load($akHandle, 'akHandle');
+        if ($ak->getAttributeKeyID() < 1) {
+            $ak = -1;
+        }
+
+        CacheLocal::set('collection_attribute_key_by_handle', $akHandle, $ak);
+
+        if ($ak === -1) {
+            return false;
+        }
+
+        return $ak;
+    }
+
+    /**
+     * Get list of attributes for collection category
+     * @param null $akCategoryHandle Ignored; included for legacy support
+     * @param null $filters Ignored; included for legacy support
+     * @return array
+     */
+    public static function getList($akCategoryHandle = null, $filters = null)
```

Comment:
@Mnkras I'd rather not use `get_called_class` and instead implement a `static::getCategoryTypeName()` on each of the categories so that extending classes and overrides don't have to worry about what their classes are named.